token-metadata/program/src/processor/metadata/print.rs

```diff
@@ -0,0 +1,194 @@
+        // if the token account is empty, we will initialize a new one but it must
+        // be an ATA account
+        assert_derivation(
+            &spl_associated_token_account::id(),
+            edition_token_account_info,
+            &[
+                edition_token_account_owner_info.key.as_ref(),
+                spl_token::id().as_ref(),
+                edition_mint_info.key.as_ref(),
+            ],
+        )?;
+
+        // creating the associated token account
+        invoke(
+            &spl_associated_token_account::instruction::create_associated_token_account(
+                ctx.accounts.payer_info.key,
+                edition_token_account_owner_info.key,
+                edition_mint_info.key,
+                &spl_token::id(),
+            ),
+            &[
+                payer_info.clone(),
+                edition_token_account_owner_info.clone(),
+                edition_mint_info.clone(),
+                edition_token_account_info.clone(),
+            ],
+        )?;
+    } else {
+        assert_owned_by(edition_token_account_info, &spl_token::id())?;
+    }
+
+    if ata_program.key != &spl_associated_token_account::ID {
+        return Err(ProgramError::IncorrectProgramId);
+    }
+
+    if sysvar_instructions.key != &sysvar::instructions::ID {
+        return Err(ProgramError::IncorrectProgramId);
+    }
+
+    // Deserialize the master edition's metadata so we can determine token type
+    let master_metadata = Metadata::from_account_info(master_metadata_info)?;
+    let token_standard = master_metadata
+        .token_standard
+        .unwrap_or(TokenStandard::NonFungible);
+
+    match token_standard {
+        TokenStandard::NonFungible => {}
+        TokenStandard::ProgrammableNonFungible => {
+            // Validate that the token record was passed in for pNFTs.
+            let token_record_info = match edition_token_record_info {
+                Some(token_record_info) => token_record_info,
+                None => return Err(MetadataError::MissingTokenRecord.into()),
+            };
+            let (pda_key, _) = find_token_record_account(
+                ctx.accounts.edition_mint_info.key,
+                ctx.accounts.edition_token_account_info.key,
+            );
+            // validates the derivation
+            assert_keys_equal(&pda_key, token_record_info.key)?;
+
+            if token_record_info.data_is_empty() {
+                create_token_record_account(
+                    &crate::ID,
+                    token_record_info,
+                    edition_mint_info,
+                    edition_token_account_info,
+                    payer_info,
+                    system_program,
+                )?;
+            } else {
+                assert_owned_by(token_record_info, &crate::ID)?;
+            }
+        }
+        _ => return Err(MetadataError::InvalidTokenStandard.into()),
+    };
+
+    process_mint_new_edition_from_master_edition_via_token_logic(
```

Comment:
It would be good to add a comment here to say that most of the accounts validation happen on the `process_mint_new_edition_from_master_edition_via_token_logic`.

Reply:
There are // CHECK comments at the top of the function